### ISSUE TITLE
NullReferenceException when calling SubscribeAsync on virtual stream (after upgrading to 1.4.1 or 1.4.2)

### ISSUE THREAD

**Issue by danvanderboom** [dotnet/orleans#3117]
I have some code that gets a virtual stream and tries to subscribe to the stream like this:
```
HeartbeatStream = StreamProvider
   .GetStream<HeartbeatEvent>(Guid.Empty, StreamNames.Heartbeats);

await HeartbeatStream.SubscribeAsync(OnNewHeartbeat);
```
This and other subscriptions have worked for a long time. After upgrading to 1.4.2, and then downgrading back to 1.4.1, this call to SubscribeAsync is now throwing an exception I haven't seen before:

> NullReferenceException
>    at Orleans.Providers.ClientProviderRuntime.ExecutingEntityIdentity()
>    at Orleans.Streams.StreamConsumerExtension.SetObserver[T](GuidId subscriptionId, StreamImpl`1 stream, IAsyncObserver`1 observer, StreamSequenceToken token, IStreamFilterPredicateWrapper filter)
>    at Orleans.Streams.StreamConsumer`1.<SubscribeAsync>d__14.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
>    at RTOrleansMon.Program.<MonitorAppVirtualStreams>d__15.MoveNext()

StreamProvider (SMS) is non-null, and HeartbeatStream is assigned to a non-null stream handle object. 

ClientProviderRuntime.ExecutingEntityIdentity is a single line:
https://github.com/dotnet/orleans/blob/master/src/Orleans/Providers/ClientProviderRuntime.cs#L103

Suggesting that runtimeClient is in some cases null?
https://github.com/dotnet/orleans/blob/master/src/Orleans/Providers/ClientProviderRuntime.cs#L26

This virtual stream subscriber is in a client... remote from the silos.

I've gone back and forth a few times for testing, both 1.4.1 and 1.4.2, ensuring that ALL of my projects are referencing the same version of all Orleans NuGets.

I upgraded from 1.3.x to 1.4.1 and saw it working after that. Remote client subscribed to virtual stream published from grains in silos. Upgraded to 1.4.2 and it broke. Downgraded back to 1.4.1 and it still throws the exception above.

Other that this, the silos themselves seem to be operating great. Data flowing through the system like it should, no errors being logged (with extensive instrumentation).

There are no signs of errors in the Orleans client logs, but the stream consumer messages that are logged look like this:
```
[2017-06-11 20:21:03.195 GMT     1	VERBOSE 	100000	StreamConsumer<HeartbeatEvent>-Heartbeats-00000000-0000-0000-0000-000000000000-SMSProvider	192.168.0.40:0]	Subscribe Observer=Orleans.Streams.GenericAsyncObserver`1[NotificationServer.Model.HeartbeatEvent] Token=	
[2017-06-11 20:21:03.199 GMT     1	VERBOSE 	100000	StreamConsumer<HeartbeatEvent>-Heartbeats-00000000-0000-0000-0000-000000000000-SMSProvider	192.168.0.40:0]	BindExtensionLazy - Binding local extension to stream runtime=Orleans.Providers.ClientProviderRuntime	
[2017-06-11 20:21:03.207 GMT     1	VERBOSE 	100298	Interner-GuidId	192.168.0.40:0]	Starting Interner-GuidId cache cleanup timer with frequency 00:10:00	
[2017-06-11 20:21:03.210 GMT     1	VERBOSE 	101404	sync.SafeTimerBase	192.168.0.40:0]	Creating timer sync.SafeTimerBase with dueTime=00:10:00 period=00:10:00	
[2017-06-11 20:21:03.214 GMT     1	VERBOSE 	100000	StreamConsumer<HeartbeatEvent>-Heartbeats-00000000-0000-0000-0000-000000000000-SMSProvider	192.168.0.40:0]	BindExtensionLazy - Connected Extension=Orleans.Streams.StreamConsumerExtension GrainRef=ObserverId:*cli/40826daa/1a3b0049-e26e-4237-a22d-996050fd4176	
```

Any ideas of where to look next?

**Comment by danvanderboom** [comment]
I tried updating to Orleans 1.5.0-beta1 and got these errors:

> System.NullReferenceException: Object reference not set to an instance of an object.  
>    at Orleans.Runtime.Messaging.IncomingMessageAcceptor.ProcessAccept(SocketAsyncEventArgs e, Boolean completedSynchronously)

> System.OutOfMemoryException: Array dimensions exceeded supported range.  
>    at Orleans.Serialization.BuiltInTypes.DeserializeInvokeMethodRequest(Type expected, IDeserializationContext context)  
>    at Orleans.Serialization.SerializationManager.DeserializeInner(Type expected, IDeserializationContext context)  
>    at Orleans.Serialization.SerializationManager.Deserialize(Type t, BinaryTokenStreamReader stream)  
>    at Orleans.Runtime.Message.DeserializeBody(SerializationManager serializationManager, List`1 bytes)

**Comment by danvanderboom** [comment]
GetConsumerInterface() returns an Orleans StreamConsumer\<T> object, and then the call to SubscribeAsync on that object throws the NullReferenceException. 

![image](https://user-images.githubusercontent.com/2482850/27042064-2659ea5a-4f53-11e7-8cd3-8af9eea5a74b.png)

Can't seem to step into the Orleans code for some reason. I downloaded the source, checked out the matching (1.4.2) tag, ...

**Comment by danvanderboom** [comment]
I reverted from 1.4.2 to 1.4.1, then to 1.4.0, and still saw the same problem on all of those. (Despite not seeing the problem for weeks after upgrading to 1.4.1.)

The only thing so far I've found that makes the problem go away is reverting to Orleans 1.3.1.

**Comment by jason-bragg** [comment]
@danvanderboom, thanks for reporting this.

Unfortunately, we've seen this bug and fixed it for 1.5 (not yet released), but failed to pull the fix into 1.4.2.  I apologize for missing this.

You are correct in your assessment that this is related to the use of ExecutingEntityIdentity, in the client.  This issue is client only and, for the most part, should be limited to environments where log level is set to Verbose or lower.

We should be releasing 1.5 soon, in the interim you can avoid this by raising the logging level.

If verbose logs are necessary, you can raise the log level only in the specific areas with this issue:  "StreamConsumerExtension" and/or "SimpleMessageStreamProducerExtension".  http://dotnet.github.io/orleans/Documentation/Orleans-Configuration-Guide/Client-Configuration.html

Hope this helps.  If this mitigation is not sufficient, please let us know.

**Comment by danvanderboom** [comment]
Great news @jason-bragg - thanks for the info about the fix being in 1.5.0. I may even squeeze in a test of that today and see what happens.

**Comment by sergeybykov** [comment]
>Can't step into the SubscribeAsync method for some reason.

Symbols for 1.4.2 were published to Microsoft Symbol Servers only yesterday (for some reason it takes a while). Also, make sure you check out the code branch for debugging. If you download it from GitHub instead, the debugger will complain about mismatch of sources because of the CR/LF difference between Windows/Unix.

**Comment by sergeybykov** [comment]
Resolved in 1.5.0.
